source/lac/scalapack.cc

```diff
@@ -761,6 +911,415 @@ NumberType ScaLAPACKMatrix<NumberType>::norm(const char type) const
   return res;
 }
 
+
+
+template <typename NumberType>
+void ScaLAPACKMatrix<NumberType>::save(const char *filename) const
+{
+#ifndef DEAL_II_WITH_HDF5
+  (void)filename;
+  AssertThrow(false, ExcMessage ("HDF5 support is disabled."));
+#else
+#  ifdef H5_HAVE_PARALLEL
+  //implementation for configurations equipped with a parallel file system
+  save_parallel(filename);
+
+#  else
+  //implementation for configurations with no parallel file system
+  save_serial(filename);
+
+#  endif
+#endif
+}
+
+
+
+template <typename NumberType>
+void ScaLAPACKMatrix<NumberType>::save_serial(const char *filename) const
+{
+#  ifndef DEAL_II_WITH_HDF5
+  (void)filename;
+  Assert(false,ExcInternalError());
+#  else
+
+  /*
+   * The content of the distributed matrix is copied to a matrix using a 1x1 process grid.
+   * Therefore, one process has all the data and can write it to a file.
+   *
+   * Create a 1x1 column grid which will be used to initialize
+   * an effectively serial ScaLAPACK matrix to gather the contents from the current object
+   */
+  std::shared_ptr<Utilities::MPI::ProcessGrid> column_grid = std::make_shared<Utilities::MPI::ProcessGrid>(this->grid->mpi_communicator,1,1);
```

Comment:
Does this pointer live beyond the end of this function? If not, could you use a `std::unique_ptr`?

Reply:
The pointer is alive only in the scope of this function. But it has to be a `std::shared_ptr` as the constructor of `ScaLAPACKMatrix`, which is called subsequently, requires it.

Reply:
p.s. @bangerth it's like `Utilities::MPI::Partitioner` for parallel vectors in dealii, we always take them as shared pointers in order to be able to share an object between matrices, if needed.

Reply:
I see. Thanks for the explanation!